plugins/duckdb/src/duckdb_session.cpp

```
@@ -0,0 +1,442 @@
+	const std::string q = normalize(sql, len);
+	if (q.empty()) return false;
+
+	static const char* const read_keywords[] = {
+		"SELECT", "WITH", "TABLE", "VALUES", "DESCRIBE", "SHOW", "PRAGMA", "EXPLAIN"
+	};
+	size_t i = 0;
+	while (i < q.size() && std::isalpha(static_cast<unsigned char>(q[i]))) i++;
+	const std::string first_word = q.substr(0, i);
+	bool starts_with_read = false;
+	for (const char* kw : read_keywords) {
+		if (first_word == kw) { starts_with_read = true; break; }
```

Cppcheck (2.21.0)

[warning] If memory allocation fails, then there is a possible null pointer dereference

(nullPointerOutOfMemory)